lapack/testlapack/matgen.go

```diff
@@ -738,3 +738,505 @@ func applyReflector(qh blas64.General, q blas64.General, v []float64) {
+				ab[i*ldab] = rnd.NormFloat64()
+				for j := 1; j < min(n-i, kd+1); j++ {
+					ab[i*ldab+j] = 0
+				}
+			}
+		} else {
+			for i := 0; i < n; i++ {
+				for j := max(0, kd-i); j < kd; j++ {
+					ab[i*ldab+j] = 0
+				}
+				// Fill the diagonal with random numbers.
+				ab[i*ldab+kd] = rnd.NormFloat64()
+			}
+		}
+
+	case 7, 8, 9:
+		// Non-trivial unit triangular matrix
+		//
+		// A unit triangular matrix T with condition cndnum is formed.
+		// In this version, T only has bandwidth 2, the rest of it is
+		// zero.
+
+		tnorm := math.Sqrt(cndnum)
+
+		// Initialize AB to zero.
+		if uplo == blas.Upper {
+			for i := 0; i < n; i++ {
+				// Fill the diagonal with random numbers.
+				ab[i*ldab] = rnd.NormFloat64()
+				for j := 1; j < min(n-i, kd+1); j++ {
+					ab[i*ldab+j] = 0
+				}
+			}
+		} else {
+			for i := 0; i < n; i++ {
+				for j := max(0, kd-i); j < kd; j++ {
+					ab[i*ldab+j] = 0
+				}
+				// Fill the diagonal with random numbers.
+				ab[i*ldab+kd] = rnd.NormFloat64()
+			}
+		}
+
+		switch kd {
+		case 0:
+			// Unit diagonal matrix, nothing else to do.
+		case 1:
+			// Special case: T is tridiagonal. Set every other
+			// off-diagonal so that the matrix has norm tnorm+1.
+			if n > 1 {
+				if uplo == blas.Upper {
+					ab[1] = math.Copysign(tnorm, uniformM11())
+					for i := 2; i < n-1; i += 2 {
+						ab[i*ldab+1] = tnorm * uniformM11()
+					}
+				} else {
+					ab[ldab] = math.Copysign(tnorm, uniformM11())
+					for i := 3; i < n; i += 2 {
+						ab[i*ldab] = tnorm * uniformM11()
+					}
+				}
+			}
+		default:
+			// Form a unit triangular matrix T with condition cndnum. T is given
+			// by
+			//      | 1   +   *                      |
+			//      |     1   +                      |
+			//  T = |         1   +   *              |
+			//      |             1   +              |
+			//      |                 1   +   *      |
+			//      |                     1   +      |
+			//      |                          . . . |
+			// Each element marked with a '*' is formed by taking the product of
+			// the adjacent elements marked with '+'. The '*'s can be chosen
+			// freely, and the '+'s are chosen so that the inverse of T will
+			// have elements of the same magnitude as T.
+			work1 := make([]float64, n)
+			work2 := make([]float64, n)
+			star1 := math.Copysign(tnorm, uniformM11())
+			sfac := math.Sqrt(tnorm)
+			plus1 := math.Copysign(sfac, uniformM11())
+			for i := 0; i < n; i += 2 {
+				work1[i] = plus1
+				work2[i] = star1
+				if i+1 == n {
+					continue
+				}
+				plus2 := star1 / plus1
+				work1[i+1] = plus2
+				plus1 = star1 / plus2
+				// Generate a new *-value with norm between sqrt(tnorm)
+				// and tnorm.
+				rexp := uniformM11()
+				if rexp < 0 {
+					star1 = -math.Pow(sfac, 1-rexp)
+				} else {
+					star1 = math.Pow(sfac, 1+rexp)
+				}
+			}
+			// Copy the diagonal to AB.
+			if uplo == blas.Upper {
+				bi.Dcopy(n-1, work1, 1, ab[1:], ldab)
```

Comment:
`n` can be 1 by here, no? Is it worth this being made conditional?

Reply:
Yes, it can. I had it conditional but then decided it's not worth it. I think it's ok to leave as is.